src/abstractdataframe/reshape.jl

```diff
@@ -399,3 +399,113 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+                dest_namescol::Union{Symbol,AbstractString};
+                makeunique::Bool=false)
+    permutedims(df::AbstractDataFrame, src_namescol::ColumnIndex; makeunique::Bool=false)
+    permutedims(df::AbstractDataFrame; makeunique::Bool=false)
+
+Turn `df` on its side such that rows become columns
+and the column indexed by `src_namescol` becomes the names of new columns.
+In the resulting `DataFrame`, column names of `df` will become the first column
+with name specified by `dest_namescol`.
+
+# Arguments
+- `df` : the `AbstractDataFrame`
+- `src_namescol` : the column that will become the new header.
+  This column eltype must be `<: Union{String, Symbol}`.
+  Defaults to first column.
+- `dest_namescol` : the name of the first column in the returned `DataFrame`.
+  Defaults to the same name as `src_namescol`.
+- `makeunique` : if `false` (the default), an error will be raised
+  if duplicate names are found; if `true`, duplicate names will be suffixed
+  with `_i` (`i` starting at 1 for the first duplicate).
+
+Note: The eltypes of columns in resulting `DataFrame`
+(other than the first column, which always has eltype `String`)
+will depend on the eltypes of _all_ input columns
+based on the results of `prote_type`.
+That is, if the source `DataFrame` contains `String` and `Int` columns,
+resulting columns will have eltype `Any`.
+If the source has a mix of numeric types (eg. `Float64` and `Int`),
+columns in resulting `DataFrame` will be promoted to `Float64`.
+
+# Examples
+
+```jldoctest
+julia> df1 = DataFrame(a=["x", "y"], b=[1.,2.], c=[3,4], d=[true,false])
+2×4 DataFrame
+│ Row │ a      │ b       │ c     │ d    │
+│     │ String │ Float64 │ Int64 │ Bool │
+├─────┼────────┼─────────┼───────┼──────┤
+│ 1   │ x      │ 1.0     │ 3     │ 1    │
+│ 2   │ y      │ 2.0     │ 4     │ 0    │
+
+julia> df2 = DataFrame(a=["x", "y"], b=[1, "two"], c=[3,4], d=[true,false])
+2×4 DataFrame
+│ Row │ a      │ b   │ c     │ d    │
+│     │ String │ Any │ Int64 │ Bool │
+├─────┼────────┼─────┼───────┼──────┤
+│ 1   │ x      │ 1   │ 3     │ 1    │
+│ 2   │ y      │ two │ 4     │ 0    │
+
+julia> permutedims(df1) # note the column types
+3×3 DataFrame
+│ Row │ a      │ x       │ y       │
+│     │ String │ Float64 │ Float64 │
+├─────┼────────┼─────────┼─────────┤
+│ 1   │ b      │ 1.0     │ 2.0     │
+│ 2   │ c      │ 3.0     │ 4.0     │
+│ 3   │ d      │ 1.0     │ 0.0     │
+
+julia> permutedims(df2)
+3×3 DataFrame
+│ Row │ a      │ x   │ y   │
+│     │ String │ Any │ Any │
+├─────┼────────┼─────┼─────┤
+│ 1   │ b      │ 1   │ two │
+│ 2   │ c      │ 3   │ 4   │
+│ 3   │ d      │ 1   │ 0   │
+```
+"""
+function Base.permutedims(df::AbstractDataFrame, src_namescol::ColumnIndex,
+                          dest_namescol::Union{Symbol, <:AbstractString};
+                          makeunique::Bool=false)
+
+    if src_namescol isa Integer
+        1 <= src_namescol <= ncol(df) || throw(BoundsError("`src_namescol` doesn't exist"))
```

Comment:
`BoundsError` does not take a string (see other `BoundsError` uses to see the syntax)

Reply:
Can I do 

```
@boundscheck if src_namescol isa Integer
        1 <= src_namescol <= ncol(df) || throw(BoundsError(names(df), src_namescol))
    end
```

?

Reply:
Now seeing examples that don't have `@boundscheck` - is that warranted here?

Reply:
No - `@boundscheck` is only needed in performance sensitive code. This function is slow, so we should always do bounds checking.

Reply:
IIRC you should do something like `BoundsError(df, src_namescol)` as we don't want to print an error about `names(df)`.

Reply:
This gives

Reply:
I guess that's clearer than

```
julia> permutedims(DataFrame(x=["a","b","c"], y=rand(3), z=rand(3)), 4)
ERROR: BoundsError: attempt to access 3-element Vector{String} at index [4]
```

Reply:
Normally we give the following error message:
```
julia> df = DataFrame(x=1)
1×1 DataFrame
│ Row │ x     │
│     │ Int64 │
├─────┼───────┤
│ 1   │ 1     │

julia> df[:, 2]
ERROR: BoundsError: attempt to access data frame with 1 column
  at index [2]
```
as I have suggested earlier - please look up the implementation there. We throw an error on `AbstractIndex` in order to generate a clean error message.

Reply:
Apologies - I had looked at a couple of other examples, but evidently not the right one. It's not super obvious how stuff fits together, but I think I have it now:

```
julia> permutedims(DataFrame(x=["a","b","c"], y=rand(3), z=rand(3)), 4)
ERROR: BoundsError: attempt to access data frame with 3 columns at index [4]
```

Reply:
Yes (and indeed DataFrames.jl is unfortunately complex, but we could not find a simpler design that would meet all the requirements people have)